torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -895,6 +911,21 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 
 void ProcessGroupNCCL::ncclCommWatchdog() {
   try {
+    // Inside watchdog thread, we might still call CUDA Event and it will
+    // hang the thread and stuck there silent for hours. We want to use a
+    // timer to kill the process if watchdog timeout.
+    struct sigevent action;
+    std::memset(&action, 0, sizeof(action));
+    action.sigev_notify = SIGEV_THREAD;
```

Comment:
not sure between `SIGEV_SIGNAL` and `SIGEV_THREAD` options

SIGNAL option could be simpler and just immediately kill our process for us, without needing to have a new thread created

THREAD option gives us a callback if we want to do additional logging or any other behavior.  Probably reliable enough i assume?

Reply:
Well, we need to delete the timer to avoid leakage. That's why I choose `SIGEV_THREAD` here.